ckanext/datastore/backend/postgres.py

```diff
@@ -934,6 +994,32 @@ def create_table(context, data_dict, plugin_data):
     sql_fields = u", ".join([u'{0} {1}'.format(
         identifier(f['id']), f['type']) for f in fields])
 
+    # (canada fork only): foreign keys
+    # TODO: upstream contrib!!
+    foreign_keys = data_dict.get('foreign_keys')
+    foreign_fields = []
+    if foreign_keys:
+        for f_table_name, column_name_map in foreign_keys.items():
+            constraint_name = _generate_constraint_name(f_table_name, column_name_map.keys(), column_name_map.values())
+            log.debug('Trying to build foreign key constraint {0} for column(s) {1}({2}) on foreign column(s) {3}({4})'.format(
+                identifier(constraint_name),
+                identifier(data_dict['resource_id']),
+                ','.join(identifier(k) for k in column_name_map.keys()),
+                identifier(f_table_name),
+                ','.join(identifier(v) for v in column_name_map.values())
+            ))
+            #TODO: allow for deletion and update enum??
+            # might be able to do this just for upstream. for now, we can force restrict delete and cascade update.
+            # a = no action, r = restrict, c = cascade, n = set null, d = set default
+            foreign_fields.append('CONSTRAINT {0} FOREIGN KEY ({1}) REFERENCES {2}({3}) ON DELETE RESTRICT ON UPDATE CASCADE'.format(
```

Comment:
update cascade might not do what we want. If the parent entries are updated in a different order (based on the `_id` values) then the codes would be replaced with the wrong ones.

This is a useful feature but we'll need to be careful about the potential for big problems if a user does this.

Reply:
going to restrict on update for now. Recombinant does not allow updating of DS primary keys, so this should not be an issue.

Need to confirm if Edit in Excel is affected at all.